### PR TITLE
Stinger done

### PR DESCRIPTION
**Pull request by Sergeanur**
Note: untested

### CODE REVIEW

File: src/objects/Stinger.cpp [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+CStinger::Remove()
+{
+	if (!bIsDeployed) return;
+
+	for (int32 i = 0; i < NUM_STINGER_SEGMENTS; i++) {
+		CStingerSegment *spikeSegment = pSpikes[i];
+		if (spikeSegment->m_entryInfoList.first != nil)
+			spikeSegment->bRemoveFromWorld = true;
+		else
+			delete spikeSegment;
```
Comment (ShFil119) [inline src/objects/Stinger.cpp:80]: `pSpikes[i]` could be nilled

File: src/objects/Stinger.cpp [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+}
+
+void
+CStinger::Init(CPed *pPed)
+{
+	int32 i;
+
+	pOwner = pPed;
+	for (i = 0; i < NUM_STINGER_SEGMENTS; i++) {
+		pSpikes[i] = new CStingerSegment;
```
Comment (ShFil119) [inline src/objects/Stinger.cpp:49]: `new CStingerSegment();`?

File: src/objects/Stinger.cpp [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+		if (pOwner != nil && pOwner->m_nPedType == PEDTYPE_COP)
+			((CCopPed*)pOwner)->m_bThrowsSpikeTrap = false;
+		break;
+	case STINGERSTATE_STATE3:
+		if (CTimer::GetTimeInMilliseconds() > m_nTimeOfDeploy + 2500)
+			m_nSpikeState = STINGERSTATE_REMOVE;
+		// no break
+	case STINGERSTATE_STATE1:
+		if (m_nSpikeState != STINGERSTATE_STATE1 || CTimer::GetTimeInMilliseconds() <= m_nTimeOfDeploy + 2500) {
+			float something = (CTimer::GetTimeInMilliseconds() - m_nTimeOfDeploy) / 2500.0f;
```
Comment (ShFil119) [inline src/objects/Stinger.cpp:191]: progress?

File: src/peds/CopPed.cpp [hunk 4]
```diff
@@ -92,6 +93,7 @@ CCopPed::CCopPed(eCopType copType, int32 modifier) : CPed(PEDTYPE_COP)
 	m_nHassleTimer = 0;
 	field_61C = 0;
 	field_624 = 0;
+	m_pStinger = new CStinger;
```
Comment (ShFil119) [inline src/peds/CopPed.cpp:96]: also `()`?

**Review by ShFil119 — APPROVED** [review]

File: src/objects/Stinger.cpp (hunk 1 above)
Comment (Sergeanur) [reply src/objects/Stinger.cpp:80]: Original code doesn't do that

File: src/objects/Stinger.cpp (hunk 2 above)
Comment (Sergeanur) [reply src/objects/Stinger.cpp:49]: nah

File: src/peds/CopPed.cpp (hunk 4 above)
Comment (Sergeanur) [reply src/peds/CopPed.cpp:96]: also nah

File: src/objects/Stinger.cpp (hunk 3 above)
Comment (Sergeanur) [reply src/objects/Stinger.cpp]: ok

Comment (aap) [reply src/objects/Stinger.cpp]: I'd flip this if. and for `something` maybe `interp` or just `t`?

File: src/objects/Stinger.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+};
+
+#define NUM_STINGER_SEGMENTS (12)
+
+class CStinger
+{
+	// TODO: better names
+	enum {
+		STINGERSTATE_NONE = 0,
+		STINGERSTATE_STATE1,
```
Comment (aap) [inline src/objects/Stinger.h:19]: 1 and 3 seem to animate the thing in opposite directions, and 2 is reached at the end of 2. So maybe EXTENDING, EXTENDED and PULLINGBACK or something similar?

File: src/objects/Stinger.cpp [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+		if (vehsInRange[i]->IsCar())
+			pAutomobile = (CAutomobile*)vehsInRange[i];
+		else if (vehsInRange[i]->IsBike())
+			pBike = (CBike*)vehsInRange[i];
+
+		if (pAutomobile == nil && pBike == nil) continue;
+
+		int wheelId = 0;
+		float someWheelDist = sq(((CVehicleModelInfo*)CModelInfo::GetModelInfo(vehsInRange[i]->GetModelIndex()))->m_wheelScale);
```
Comment (erorcun) [inline src/objects/Stinger.cpp:125]: not a good variable name

File: src/objects/Stinger.cpp [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+CStinger::CheckForBurstTyres()
+{
+	const CVector firstPos = pSpikes[0]->GetPosition();
+	const CVector lastPos = pSpikes[NUM_STINGER_SEGMENTS - 1]->GetPosition();
+	float dist = (lastPos - firstPos).Magnitude();
+	if (dist > 0.1f) return;
+
+	CVehicle *vehsInRange[16];
+	int16 numObjects;
+	CEntity someEntity;
```
Comment (erorcun) [inline src/objects/Stinger.cpp:107]: this is unused (and not a good name anyway)

File: src/objects/Stinger.cpp [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+			CColPoint colPoint;
+			CEntity *pEntity;
+			if (CWorld::ProcessVerticalLine(CVector(pos3d.x, pos3d.y, pos3d.z - 10.0f), pos3d.z, colPoint, pEntity, true, false, false, false, true, false, nil))
+				pos3d.z = colPoint.point.z + 0.15f;
+
+			angle1 = CGeneral::LimitRadianAngle(angle1);
+			angle2 = CGeneral::LimitRadianAngle(angle2);
+
+			for (int spike = 0; spike < NUM_STINGER_SEGMENTS; spike++) {
+				CVector somePosAgain = pos3d + CVector(pos2d.x, pos2d.y, 0.6f);
```
Comment (erorcun) [inline src/objects/Stinger.cpp:213]: again

**Pull request merged**